service/resource/project/resource.go

```diff
@@ -73,105 +90,140 @@ func (r *Resource) GetDesiredState(obj interface{}) (interface{}, error) {
+	var projectsToDelete []Project
 
-	r.logger.Log("debug", fmt.Sprintf("found delete state: %#v", nil))
+	for _, currentProject := range currentProjects {
+		if existsProjectByName(desiredProjects, currentProject.Name) {
+			projectsToDelete = append(projectsToDelete, currentProject)
+		}
+	}
 
-	return nil, nil
+	return projectsToDelete, nil
 }
 
 func (r *Resource) GetUpdateState(obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
-	customObject, err := toCustomObject(obj)
+	currentProjects, err := toProjects(currentState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	desiredProjects, err := toProjects(desiredState)
 	if err != nil {
 		return nil, nil, nil, microerror.Mask(err)
 	}
 
-	r.logger.Log("debug", "get delete state")
+	var projectsToUpdate []Project
 
-	r.logger.Log("TODO", fmt.Sprintf("implement logic based on received custom object: %#v", customObject))
+	for _, desiredProject := range desiredProjects {
+		if !existsProjectByName(currentProjects, desiredProject.Name) {
+			continue
+		}
 
-	r.logger.Log("debug", fmt.Sprintf("found delete state: %#v", nil))
+		currentProject, err := getProjectByName(currentProjects, desiredProject.Name)
+		if err != nil {
+			return nil, nil, nil, microerror.Mask(err)
+		}
+
+		if currentProject.Ref != desiredProject.Ref {
+			projectsToUpdate = append(projectsToUpdate, desiredProject)
+		}
+	}
 
-	return nil, nil, nil, nil
+	return []Project{}, []Project{}, projectsToUpdate, nil
 }
 
 func (r *Resource) Name() string {
 	return Name
 }
 
 func (r *Resource) ProcessCreateState(obj, createState interface{}) error {
-	customObject, err := toCustomObject(obj)
+	projectsToCreate, err := toProjects(createState)
 	if err != nil {
 		return microerror.Mask(err)
 	}
 
-	r.logger.Log("debug", "process create state")
+	if len(projectsToCreate) != 0 {
+		r.logger.Log("debug", "creating projects in the Kubernetes cluster")
 
-	r.logger.Log("TODO", fmt.Sprintf("implement logic based on received custom object: %#v", customObject))
+		for _, p := range projectsToCreate {
+			err := r.installer.Install(installerspec.Project{Name: p.Name, Ref: p.Ref})
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
 
-	r.logger.Log("debug", "processed create state")
+		r.logger.Log("debug", "created projects in the Kubernetes cluster")
+	} else {
+		r.logger.Log("debug", "the projects are already created in the Kubernetes cluster")
+	}
 
 	return nil
 }
 
 func (r *Resource) ProcessDeleteState(obj, deleteState interface{}) error {
-	customObject, err := toCustomObject(obj)
-	if err != nil {
-		return microerror.Mask(err)
-	}
-
-	r.logger.Log("debug", "process delete state")
-
-	r.logger.Log("TODO", fmt.Sprintf("implement logic based on received custom object: %#v", customObject))
-
-	r.logger.Log("debug", "processed delete state")
-
+	r.logger.Log("TODO", "implement ProcessDeleteState")
```

Comment:
Right now we do not handle the deletion. We have everything prepared for this though. Once we want this we only have to work on `GetCurrentState` to store the formerly received projects in the custom objects. Once we consider the projects which are removed from the TPO but are still running within the k8s cluster and add them to the current state, the deletion here can do its job. We then only need `Installer.Delete(installerspec.Project) error`.